### PR TITLE
Initial http source

### LINKED ISSUE #308: Basic log HTTP source plugin
We will add a log-http-source Source plugin. This issue tracks the following features

* Boiler plate plugin code
* Add basic http server
* Add the http service that parse the incoming http request data and push to buffer

### PR DESCRIPTION
**Pull request by chenqi0805**
### Description
This PR adds the starter for log-http-source Source plugin. It includes

- Implementing the start() method: Build armeria server, load the log ingestion service, start the server
- Implementing the stop() method: stop the server
- Add Codec Interface and implement JsonCodec for parsing http/json request data
- Add LogThrottlingStrategy and ThrottlingRejectionHandler for future loading of ThrottlingService
- Add initial README as starter

Besides automated unit tests, the source plugin has been smoke tested with FluentBit HTTP output:

```
Fluent Bit v1.8.4
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/09/22 16:40:06] [ info] [engine] started (pid=1)
[2021/09/22 16:40:06] [ info] [storage] version=1.1.1, initializing...
[2021/09/22 16:40:06] [ info] [storage] in-memory
[2021/09/22 16:40:06] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/09/22 16:40:06] [ info] [cmetrics] version=0.2.1
[2021/09/22 16:40:06] [ info] [sp] stream processor started
[2021/09/22 16:40:11] [ info] [output:http:http.0] host.docker.internal:2021, HTTP status=200
200 OK
[2021/09/22 16:40:15] [ info] [output:http:http.0] host.docker.internal:2021, HTTP status=200
200 OK
[2021/09/22 16:40:30] [error] [output:http:http.0] host.docker.internal:2021, HTTP status=408
Pipeline [test-pipeline] - Buffer is full, timed out waiting for a slot
[2021/09/22 16:40:30] [ warn] [engine] failed to flush chunk '1-1632328815.902484900.flb', retry in 6 seconds: task_id=0, input=dummy.0 > output=http.0 (out_id=0)
[2021/09/22 16:40:35] [error] [output:http:http.0] host.docker.internal:2021, HTTP status=408
Pipeline [test-pipeline] - Buffer is full, timed out waiting for a slot
[2021/09/22 16:40:35] [ warn] [engine] failed to flush chunk '1-1632328820.867386700.flb', retry in 6 seconds: task_id=1, input=dummy.0 > output=http.0 (out_id=0)
[2021/09/22 16:40:40] [error] [output:http:http.0] host.docker.internal:2021, HTTP status=408
Pipeline [test-pipeline] - Buffer is full, timed out waiting for a slot
```
 
### Issues Resolved
#308 
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPSource.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import com.linecorp.armeria.server.Server;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.server.throttling.ThrottlingService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+
+@DataPrepperPlugin(name = "otel_trace_source", type = PluginType.SOURCE)
```
Comment (graytaylor0) [inline data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPSource.java]: Change name from otel_trace_source

Comment (chenqi0805) [reply data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPSource.java]: Should have been changed. You might be looking at an outdated commit.

File: data-prepper-plugins/log-http-source/README.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# Log HTTP Source
+
+This is a source plugin that supports HTTP protocol. Currently ONLY support Json UTF-8 codec for incoming request.
```
Comment (cmanning09) [inline data-prepper-plugins/log-http-source/README.md:3]: Its more specific than just JSON UTF-8 right? This plugin only supports JSON arrays for the body of the incoming requests.

**Review by cmanning09 — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPSourceConfig.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        this.maxPendingRequests = maxPendingRequests;
+    }
+
+    public static LogHTTPSourceConfig buildConfig(final PluginSetting pluginSetting) {
+        return new LogHTTPSourceConfig(
+                pluginSetting.getIntegerOrDefault(PORT, DEFAULT_PORT),
+                pluginSetting.getIntegerOrDefault(REQUEST_TIMEOUT, DEFAULT_REQUEST_TIMEOUT_MS),
+                pluginSetting.getIntegerOrDefault(THREAD_COUNT, DEFAULT_THREAD_COUNT),
+                pluginSetting.getIntegerOrDefault(MAX_CONNECTION_COUNT, DEFAULT_MAX_CONNECTION_COUNT),
+                pluginSetting.getIntegerOrDefault(MAX_PENDING_REQUESTS, DEFAULT_MAX_PENDING_REQUESTS)
```
[the remark this answers is not in the document]
Comment (dinujoh) [reply data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPSourceConfig.java:52]: We need to add the validation in the constructor to fail fast.

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPService.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        List<String> jsonList;
+        try {
+            jsonList = jsonCodec.parse(aggregatedHttpRequest.content());
+        } catch (IOException e) {
+            return HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.ANY_TYPE, "Bad request data format. Needs to be json array.");
+        }
+        for (String json: jsonList) {
+            try {
+                // TODO: switch to new API writeAll once ready
+                buffer.write(new Record<>(json), bufferWriteTimeoutInMillis);
```
Comment (dinujoh) [inline data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPService.java:60]: Do we want to write partial data into buffer ? If not, then the buffer write should take list of Records, acquire the number of permits equal to Record List size and then write to buffer else throw TimeoutException.


```tryAcquire(int permits, long timeout, TimeUnit unit)```

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingRejectHandler.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.throttling.ThrottlingRejectHandler;
+
+import javax.annotation.Nullable;
+
+public class LogThrottlingRejectHandler implements ThrottlingRejectHandler<HttpRequest, HttpResponse> {
+    private final int maxPendingRequests;
+
+    public LogThrottlingRejectHandler(int maxPendingRequests) {
+        this.maxPendingRequests = maxPendingRequests;
```
Comment (dinujoh) [inline data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingRejectHandler.java:28]: Add validation here.

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingStrategy.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletionStage;
+
+public class LogThrottlingStrategy extends ThrottlingStrategy<HttpRequest> {
+    private final int maxPendingRequests;
+    private final BlockingQueue<Runnable> queue;
+
+    public LogThrottlingStrategy(int maxPendingRequests, BlockingQueue<Runnable> queue) {
+        this.maxPendingRequests = maxPendingRequests;
+        this.queue = queue;
```
Comment (dinujoh) [inline data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingStrategy.java:28]: add validation

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPService.java (hunk 4 above)
Comment (chenqi0805) [reply data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPService.java:60]: So this change needs to happen at the buffer API level rather than here in the service handler. Thus I made the TODO comment. Right now it is accepting partial data, but it will be fixed in the buffer::writeAll.

File: data-prepper-plugins/log-http-source/README.md (hunk 2 above)
Comment (chenqi0805) [reply data-prepper-plugins/log-http-source/README.md:3]: So for pure json e.g. {"a": "b"}, it is called `json_lines` in fluentbit format and logstash codec plugin, I will add example in-line to clarify what json codec means

File: data-prepper-plugins/log-http-source/README.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+* thread_count(Optional) => the number of threads to keep in the ScheduledThreadPool. Default is `200`.
+* max_connection_count(Optional) => the maximum allowed number of open connections. Default is `500`.
+* max_pending_requests(Optional) => the maximum allowed number of tasks in the ScheduledThreadPool work queue. Default is `1024`.
+
+## Metrics
+
+TBD
+
+## Developer Guide
+This plugin is compatible with Java 8. See
```
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-plugins/log-http-source/README.md:26]: Sure. This is kind of a minimum version. But I will bump it to Java 14 for alignment.

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPService.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    private final JsonCodec jsonCodec = new JsonCodec();
+    private final Buffer<Record<String>> buffer;
+    private final int bufferWriteTimeoutInMillis;
+
+    public LogHTTPService(int bufferWriteTimeoutInMillis, Buffer<Record<String>> buffer) {
+        this.buffer = buffer;
+        this.bufferWriteTimeoutInMillis = bufferWriteTimeoutInMillis;
+    }
+
+    @Get("/log/ingest")
```
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPService.java:40]: Just to give client more options in making HTTP request, usually request with data is sent through POST method.

File: data-prepper-plugins/log-http-source/src/test/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        assertEquals(HttpStatus.BAD_REQUEST, postResponse.status());
+    }
+
+    @Test
+    public void testHTTPRequestTimeout() throws InterruptedException, ExecutionException, JsonProcessingException {
+        // Prepare
+        AggregatedHttpRequest populateDataRequest = generateRandomValidHTTPRequest(HttpMethod.GET, 2);
+        AggregatedHttpResponse getResponse = logHTTPService.doGet(populateDataRequest).aggregate().get();
+        assertEquals(HttpStatus.OK, getResponse.status());
+        AggregatedHttpRequest timeoutRequest = generateRandomValidHTTPRequest(HttpMethod.GET, 2);
```
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-plugins/log-http-source/src/test/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPServiceTest.java:96]: Sure. I can add that.

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPSourceConfig.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPSourceConfig.java:52]: I will add validation in the constructor.

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingRejectHandler.java (hunk 5 above)
Comment (chenqi0805) [reply data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingRejectHandler.java:28]: I will add validation in the constructor of LogHTTPSourceConfig since that is the entry point for input

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingStrategy.java (hunk 6 above)
Comment (chenqi0805) [reply data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingStrategy.java:28]: What do we want to validate on queue?

Comment (dinujoh) [reply data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingStrategy.java:28]: not null check for queue and min and max value validation for maxPendingRequests

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingRejectHandler.java (hunk 5 above)
Comment (dinujoh) [reply data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingRejectHandler.java:28]: It's good to have validation here if we plan to decouple with LogHTTPSourceConfig or use this with other class.

File: data-prepper-plugins/log-http-source/README.md [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    - log_http_source:
+```
+
+## Configurations
+
+* port(Optional) => An `int` represents the port source is running on. Default is ```2021```.
+* request_timeout(Optional) => An `int` represents request timeout in millis. Default is ```10_000```. 
+* thread_count(Optional) => the number of threads to keep in the ScheduledThreadPool. Default is `200`.
+* max_connection_count(Optional) => the maximum allowed number of open connections. Default is `500`.
+* max_pending_requests(Optional) => the maximum allowed number of tasks in the ScheduledThreadPool work queue. Default is `1024`.
```
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-plugins/log-http-source/README.md:19]: Good catch. I will specify the valid range for all.

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPService.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (cmanning09) [reply data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPService.java:40]: Do users want this?

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingRejectHandler.java (hunk 5 above)
Comment (cmanning09) [reply data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingRejectHandler.java:28]: This will create duplicate checks throughout our repository. I am all for being defensive but redundant code is hard to maintain. I think we need to come to an agreement and define expectations as a group.

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPService.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPService.java:40]: For Fluentbit we only need GET method. It should not hurt to expose an additional HTTP method as they all invoke the exact same business logic. But I could remove POST at the moment.

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingRejectHandler.java (hunk 5 above)
Comment (chenqi0805) [reply data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingRejectHandler.java:28]: > This will create duplicate checks throughout our repository. I am all for being defensive but redundant code is hard to maintain. I think we need to come to an agreement and define expectations as a group.

+1.

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingStrategy.java (hunk 6 above)
Comment (chenqi0805) [reply data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingStrategy.java:28]: Will add nonnull check as a safeguard. But for validation on maxPendingRequests we do need an agreement as mentioned above.

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPSource.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+import com.amazon.dataprepper.model.source.Source;
+import com.linecorp.armeria.server.Server;
+import com.linecorp.armeria.server.ServerBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+
+@DataPrepperPlugin(name = "log_http_source", type = PluginType.SOURCE)
```
Comment (dlvenable) [inline data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPSource.java:28]: This name should not contain `_source`. It doesn't add any value to somebody configuring a pipeline.

Additionally, does this work beyond just logs? Perhaps we can call this plugin just `http`?

The current `file` source is a good example to follow for this plugin.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingStrategy.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    public LogThrottlingStrategy(final int maxPendingRequests, final BlockingQueue<Runnable> queue) {
+        Objects.requireNonNull(queue);
+        this.maxPendingRequests = maxPendingRequests;
+        this.queue = queue;
+    }
+
+    @Override
+    public CompletionStage<Boolean> accept(ServiceRequestContext ctx, HttpRequest request) {
+        /*
+         * FIXME:
```
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogThrottlingStrategy.java:36]: Will do

File: data-prepper-plugins/log-http-source/src/test/java/com/amazon/dataprepper/plugins/source/loghttp/codec/JsonCodecTest.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+import com.linecorp.armeria.common.HttpData;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+class JsonCodecTest {
```
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-plugins/log-http-source/src/test/java/com/amazon/dataprepper/plugins/source/loghttp/codec/JsonCodecTest.java:23]: I will add more bad data tests that applies to our use case. Exhausting all corner cases is beyond our scenario and focused more on testing the robustness of Jackson itself.

File: data-prepper-plugins/log-http-source/src/test/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPSourceConfigTest.java [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        // When/Then
+        assertEquals(LogHTTPSourceConfig.DEFAULT_PORT, logHTTPSourceConfig.getPort());
+        assertEquals(LogHTTPSourceConfig.DEFAULT_REQUEST_TIMEOUT_MS, logHTTPSourceConfig.getRequestTimeoutInMillis());
+        assertEquals(LogHTTPSourceConfig.DEFAULT_THREAD_COUNT, logHTTPSourceConfig.getThreadCount());
+        assertEquals(LogHTTPSourceConfig.DEFAULT_MAX_CONNECTION_COUNT, logHTTPSourceConfig.getMaxConnectionCount());
+        assertEquals(LogHTTPSourceConfig.DEFAULT_MAX_PENDING_REQUESTS, logHTTPSourceConfig.getMaxPendingRequests());
+    }
+
+    @Test
+    public void testValidConfig() {
```
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-plugins/log-http-source/src/test/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPSourceConfigTest.java:46]: I suppose you mean testInvalidConfig? For valid config we just need to verify the returned attribute is the same as input

File: data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPSource.java (hunk 11 above)
Comment (chenqi0805) [reply data-prepper-plugins/log-http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/LogHTTPSource.java]: Renamed main classes correspondingly except for LogHttpService since it is not yet expected to work under generic data.

**Review by cmanning09 — APPROVED** [review]

**Review by dlvenable — APPROVED** [review]
